### PR TITLE
DI fixes for resilience

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <https://bugs.swift.org/browse/SR-5649>, <rdar://problem/33767516>, <rdar://problem/33137910>.

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp]: Please do not do this! If you truly want a SingleValueInstruction, dyn_cast to it. That being said, here you should be handling multiple results.

**Review by gottesmm — DISMISSED** [review]
Metacomment: Please do not commit the less specific error messages (I realize this is WIP).

**Comment by slavapestov** [comment]
Thanks for the suggestion, I'll definitely fix up the multiple values usages. And of course the PR will get proper commit messages and a lot of cleanup before I merge.

**Comment by slavapestov** [comment]
Still needs tests, and I have more diagnostics improvements pending.

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
Hey @jrose-apple @gottesmm this is ready for review.

There is a regression in diagnostics where sometimes we now point to the end location of a function instead of the 'return' in question when you return without initializing self. Technically this isn't a new problem, it was there all along, because the old logic was only happening with root initializers. Now that more initializers are delegating it's happening more.

Do you mind if I fix it in a follow-up PR? @jrose-apple needs the changes here for some resilience work he's doing.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test Linux

File: include/swift/AST/DiagnosticsSIL.def
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSIL.def]: Nitpick: why have three separate messages here, when it's still a choice above?

File: test/Interpreter/protocol_extensions.swift
Comment (jrose-apple) [inline test/Interpreter/protocol_extensions.swift]: Once upon a time these were deliberately minimal tests, where if something broke in StdlibUnittest it wouldn't automatically break in these tests. Maybe that's not so relevant now, though.

File: test/decl/init/resilience.swift
Comment (jrose-apple) [inline test/decl/init/resilience.swift]: What message do we get at the moment? It's worth including just so we can see when it changes.

**Review by jrose-apple — COMMENTED** [review]
Thanks, Slava!

I'd like someone more familiar with DI to review the DI changes. I'm also unhappy with the bad diagnostics not being put in the file; we should put them in in their bad state and change them later.

File: test/decl/init/resilience.swift
Comment (slavapestov) [reply test/decl/init/resilience.swift]: We now get a Sema error saying that 'self.h' is a let property. The initializer is now delegating since it's inlinable, so we don't even get as far as DI here, instead we hit the restriction that delegating inits cannot assign to let properties at all.

This isn't the "bad diagnostic" case I was talking about; that's in the DI tests (look at the patch that makes all enum constructors delegating).

File: test/Interpreter/protocol_extensions.swift
Comment (slavapestov) [reply test/Interpreter/protocol_extensions.swift]: Yeah, I understand the reasoning here, but I find StdlibUnittest easier to work with when a test fails, and also CHECK: lines are not totally reliable because it's easy to accidentally match the wrong thing. At one point Dmitri told me that new executable tests should use StdlibUnittest so I've been slowly converting them when I touch them.

File: include/swift/AST/DiagnosticsSIL.def
Comment (slavapestov) [reply include/swift/AST/DiagnosticsSIL.def]: Because the choice was keyed purely on whether its a delegating init or not, but I also want to distinguish the value type case so that diagnostics say that an assignment to self is OK too. I'll probably rework the other diagnostics when I do the fix for the return location analysis.

If you think they should still be a single diagnostic with a %select, I don't mind doing that. It seemed a bit silly to define a new enum that's only used in one place, but if I see other diagnostics get too repetitive I can do it.

File: test/decl/init/resilience.swift
Comment (jrose-apple) [reply test/decl/init/resilience.swift]: That should still be included, though. We can fix it later.

Comment (jrose-apple) [reply test/decl/init/resilience.swift]: The enum thing does seem like something we can fix in a follow-up PR. We just need to not ship it!

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Pull request merged**
